### PR TITLE
[FIX] mail: Send message/log note in chatter style with MILK

### PR DESCRIPTION
**Pull request by alexkuhn**
[FIX] mail: Send message/log note in chatter style with MILK

- Send button has text "Send" or "Log" depending on composer mode
- When button is active, it has `btn-primary` and composer mode buttons use `btn-outline-primary`
- Add hint to use Send keyboard shortcut for fast send in chatter

Other changes:

- Re-introduce missing feature to jump to present when posting a new message in a thread.
- In chatter, the topbar and composer are sticky. Whole chatter is scrollable rather than just a part of it. As a result, `Chatter-scrollable` becomes `Chatter-content`.
- Editing a message in chatter now uses the extended composer

Task-3356271

https://github.com/odoo/enterprise/pull/42111

Screenshot 2023-06-07 at 12 03 27
Screenshot 2023-06-07 at 12 03 17
Screenshot 2023-06-07 at 12 03 21
message
note

### CODE REVIEW

**Review by stefanorigano — COMMENTED** [review]
Nice @alexkuhn 👍 
Two minor remarks and a question, thanks!

File: addons/mail/static/src/web/chatter.xml [hunk 1]
```diff
@@ -89,8 +98,8 @@
             <t t-else="" t-set="placeholder">Log an internal note...</t>
             <t t-set="type" t-value="state.thread.composer.type === 'message' ? 'message' : 'note'"/>
             <SuggestedRecipientsList t-if="props.hasFollowers and state.thread.composer.type !== 'note'" className="'px-3'" styleString="'margin-left:48px;'" thread="state.thread"/>
-            <Composer composer="state.thread.composer" autofocus="true" mode="'extended'" onPostCallback.bind="onPostCallback" placeholder="placeholder" dropzoneRef="rootRef" className="state.thread.composer.type === 'message' ? 'pt-0' : ''"/>
-            <hr/>
+            <Composer composer="state.thread.composer" autofocus="true" mode="'extended'" onPostCallback.bind="onPostCallback" placeholder="placeholder" dropzoneRef="rootRef"/>
+            <hr class="mb-0"/>
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/web/chatter.xml:102]: So no `<hr>` altogether?

Only rationale for keeping `<hr>` is so that we see the top of scrollable message list below.
Otherwise we can remove it if you think it's best without it

File: addons/mail/static/src/composer/composer.xml [hunk 2]
```diff
@@ -72,15 +74,18 @@
...
                             </FileUploader>
                         </div>
                         <button t-if="thread and thread.type === 'chatter'" class="btn fa fa-expand mx-1 border-0 rounded-pill" title="Full composer" aria-label="Full composer" type="button" t-on-click="onClickFullComposer"/>
                     </div>
                     <t t-if="!extended" t-call="mail.Composer.sendButton"/>
                 </div>
-                <t t-if="extended" t-call="mail.Composer.sendButton"/>
+                <div t-if="extended" class="d-flex align-items-center mt-2">
+                    <t t-call="mail.Composer.sendButton"/>
+                    <span t-if="!props.composer.isSendButtonDisabled" class="ms-3" t-out="CTRL_ENTER_TO_SEND"/>
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/composer/composer.xml:87]: I didn't know `gap` exist!

File: addons/mail/static/src/web/chatter.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/mail/static/src/web/chatter.xml:102]: Indeed, I did not get it at first.
I think that the main problem is that it's not consistent, eg when the composer is collapsed.

https://github.com/odoo/odoo/assets/8558351/3b96e402-445b-4763-8796-2d986f1b5ebf

.. but even in that case, I think that a line (or a shadow) should appear when needed only. I made a brief POC, do you think that it would be possible to implement something similar for 16.3?

cc @paju-odoo , @seb-odoo 



https://github.com/odoo/odoo/assets/8558351/cffeee00-5fcd-485b-b0ca-a6ef4c8bcb82

[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/mail/static/src/web/chatter.xml:102]: Works better with `border-bottom`

https://github.com/odoo/odoo/assets/8558351/afa134f0-4820-40f1-a2f7-96471bf14efe

for reference, the crappy code:

```js
$(".o-mail-Chatter-scrollable").on('scroll', function() {
    $('.o-mail-Composer').toggleClass('border-bottom', $(".o-mail-Chatter-scrollable").scrollTop() !=0)
});
```

[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/web/chatter.xml:102]: `shadow-sm` is the best IMHO: `border-bottom` is too similar to date separator, `shadow` is too much.

This is how it looks: (1st is aside, 2nd is non-aside, when in bg-sheet below)
https://github.com/odoo/odoo/assets/6569390/40e6e65a-afb0-4dcf-a106-0f51fc7104b6
https://github.com/odoo/odoo/assets/6569390/90bbb989-e86b-44d5-9f19-b7c7f16ca78c

_(At the time of writing, the branch is updated but it's not shown on PR because there's a Github outage on PR specifically)_

[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/mail/static/src/web/chatter.xml]: Hei, thanks, it works flawlessly 🙌
Some notes:

- I'd suggest to `position-md-sticky` rather than `position-sticky`. Reason is that when writing a long message on mobile the rest of the chatter is not readable and the composed can't be collapsed".

https://github.com/odoo/odoo/assets/8558351/cbf867b9-5f9c-4338-9e89-12c0d87cb3db

- I think we can slightly improve the design in order to make the shadow feel more natural and to deemphasize the resulting gap when the formView hasn't scrollbars.
I'd suggest adding:

```scss
.o-mail-Chatter-top.shadow-sm {
    background-image: linear-gradient(90deg, transparent, $o-view-background-color, transparent); 
}

```

[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/web/chatter.xml]: I made the changes, thanks!!

**Comment by paju-odoo** [comment]
Hi @stefanorigano @alexkuhn,
Sorry to answering only now... 
I must say I really like this change! It looks flawless :) 
I was wondering if we could change the "edit" behavior to make it more consistent: ctrl-enter to save, instead of enter ?
Furthermore, most users are not aware of "shift-enter" to add a line break and end up sending half-written messages

**Comment by paju-odoo** [comment]
> 

Thanks! 

`Furthermore, most users are not aware of "shift-enter" to add a line break and end up sending half-written messages`
Sorry for the misunderstanding, I didn't mean to show it to users... I just think it's another argument for replacing "_enter_" with "_ctrl-enter_" in the chatter, so that the message isn't posted too easily, without it to be the user's intention :) 
That, plus the consistency with the note creation.

In discuss and chat window, that's a different behavior and I would keep _enter_ to post the message, as user are used to it. Furthermore, it's not that risky since it's a private message... You won't risk posting a half-finished message to the whole company. 

In conclusion, all good, we do agee :)

**Comment by alexkuhn** [comment]
I made the change, so now it shows the exact same composer visual in chatter, whether from composing a new message or editing a message:

Screenshot 2023-06-09 at 18 56 01

This is editing a log note. No edit on sent message (why not allow it? @paju-odoo). Visually would be clearer if logged notes had bubble layout too, e.g. gray bubble. But we can keep this improvement (or discussions about it) for another time ;)

File: addons/mail/static/src/composer/composer.xml [hunk 3]
```diff
@@ -90,20 +96,22 @@
...
         t-on-click="sendMessage"
         type="button"
-        t-att-disabled="isSendButtonDisabled"
+        t-att-disabled="props.composer.isSendButtonDisabled"
         aria-label="Send"
     >
-        <t t-if="store.isSmall"><i class="fa fa-paper-plane-o"/></t>
+        <t t-if="thread and thread.type === 'chatter' and props.composer.type === 'message'">Send</t>
+        <t t-elif="thread and thread.type === 'chatter' and props.composer.type === 'note'">Log</t>
+        <t t-elif="store.isSmall"><i class="fa fa-paper-plane-o"/></t>
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/composer/composer.xml]: done (I made the change for whole `Composer`)

File: addons/mail/static/src/web/chatter.scss [hunk 4]
```diff
@@ -10,6 +10,16 @@
     width: $o-statusbar-height;
 }
 
+.o-mail-Chatter-top {
+    top: -1px; // necessary for useStickyPinned() to detect pinned state
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/web/chatter.scss:14]: Indeed, but that's the trick used to detect when state becomes pinned, to show the gradient separator.
I will find another way to make this work, e.g. plain `t-on-scroll`

File: addons/mail/static/src/core_ui/thread.js [hunk 5]
```diff
@@ -33,6 +34,7 @@ export class Thread extends Component {
     static props = [
         "isInChatWindow?",
         "hasScrollAdjust?",
+        "jumpPresent?",
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/core_ui/thread.js:37]: In Discuss code yes it's always used, because we have a composer all the time. But I think it's good practice to reduce the amount of mandatory props to the minimum, so components are easier to reuse.

Here it makes sense to not make handling of `jumpPresent` mandatory. It's not a programming error, just a convenience for reusing `<Thread>`

File: addons/mail/static/src/composer/composer.js [hunk 6]
```diff
@@ -175,6 +193,35 @@ export class Composer extends Component {
...
+        if (this.props.composer.message && ev.target.dataset?.type === EDIT_CLICK_TYPE.SAVE) {
+            this.editMessage(ev);
+        }
+    }
+
+    get CANCEL_OR_SAVE_EDIT_TEXT() {
+        return markup(
+            sprintf(
+                _t(
+                    "<samp>%(cancel_keybind)s</samp><i> to <a href='#' data-type='%(cancel_type)s'>cancel</a></i>, <samp>%(save_keybind)s</samp><i> to <a href='#' data-type='%(save_type)s'>save</a></i>"
```
Comment (seb-odoo) [inline addons/mail/static/src/composer/composer.js:209]: `btn-link` instead of fake href with # and prevent?

File: addons/mail/static/src/composer/composer.js [hunk 7]
```diff
@@ -187,15 +234,6 @@ export class Composer extends Component {
...
     }
 
-    get isSendButtonDisabled() {
-        const attachments = this.props.composer.attachments;
-        return (
-            !this.state.active ||
-            (!this.props.composer.textInputContent && attachments.length === 0) ||
-            attachments.some(({ uploading }) => Boolean(uploading))
-        );
-    }
```
Comment (seb-odoo) [inline addons/mail/static/src/composer/composer.js:197]: why move?

File: addons/mail/static/src/composer/composer_model.js [hunk 8]
```diff
@@ -1,6 +1,7 @@
 /* @odoo-module */
 
 export class Composer {
+    active = true;
```
Comment (seb-odoo) [inline addons/mail/static/src/composer/composer_model.js:4]: could stay in component?

**Comment by seb-odoo** [comment]
@robodoo override=ci/security

`markup` are built statically

File: addons/mail/static/src/core_ui/thread.xml [hunk 9]
```diff
@@ -90,7 +90,7 @@
 </t>
 
 <t t-name="mail.Thread.jumpPresent" owl="1">
-    <span t-if="state.showJumpPresent" class="o-mail-Thread-jumpPresent position-sticky btn btn-link alert alert-info d-flex cursor-pointer align-items-center py-2 m-0" t-att-class="{ 'px-4': !env.inChatWindow, 'px-2': env.inChatWindow, 'top-0': props.order !== 'asc', 'bottom-0': props.order === 'asc' }" role="button" t-on-click="onClickJumpPresent">
+    <span t-if="state.showJumpPresent" class="o-mail-Thread-jumpPresent position-sticky btn btn-link alert alert-info d-flex cursor-pointer align-items-center py-2 m-0" t-att-class="{ 'px-4': !env.inChatWindow, 'px-2': env.inChatWindow, 'top-0': props.order !== 'asc', 'bottom-0': props.order === 'asc' }" role="button" t-on-click="() => this.jumpToPresent()">
```
Comment (seb-odoo) [inline addons/mail/static/src/core_ui/thread.xml:93]: need of arrow function?

**Review by seb-odoo — APPROVED** [review]
@robodoo r+
